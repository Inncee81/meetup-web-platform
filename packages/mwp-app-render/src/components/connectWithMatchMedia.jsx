//@flow
import * as React from 'react';
import { connect } from 'react-redux';
import type { MapStateToProps, ConnectedComponentClass } from 'react-redux';
import { createStructuredSelector, createSelector } from 'reselect';
import withMatchMedia from 'meetup-web-components/lib/utils/components/withMatchMedia';

const mapStateToProps: MapStateToProps<*, *, *> = (state: MWPState) => ({
	media: state.config.media,
});

/**
 * connectWithMatchMediaComponent
 * HOC that provides an `initialMedia` prop based on `media` in state.config which is determined by user agent
 * because withMatchMedia cannot detect the media until after initial render.
 *
 * `media` prop is provided from redux state in config.
 */
const connectWithMatchMedia = <Props: {}>(
	WrappedComponent: React.ComponentType<Props>,
): ConnectedComponentClass<*,$Diff<{ initialMedia: MatchMedia }, Props>> => {
	const MediaWrappedComponent = withMatchMedia(WrappedComponent);
	const ConnectWithMatchMedia = (props: $Diff<{ media: MatchMedia }, Props>) => (
		<MediaWrappedComponent {...props} initialMedia={props.media} />
	);

	const wrappedComponentName =
		WrappedComponent.displayName || WrappedComponent.name || 'Component';

	ConnectWithMatchMedia.displayName = `ConnectWithMatchMedia(${wrappedComponentName})`;

	return connect(mapStateToProps)(ConnectWithMatchMedia);
};

export default connectWithMatchMedia;
