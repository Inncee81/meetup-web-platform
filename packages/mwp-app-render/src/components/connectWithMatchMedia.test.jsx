import jsdom from 'jsdom';
const doc = jsdom.jsdom('<!doctype html><html><body></body></html>');
global.document = doc;
global.window = doc.defaultView;

import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import connectWithMatchMedia from './connectWithMatchMedia';

export const createFakeStore = fakeData => ({
	getState() {
		return fakeData;
	},
	dispatch() {},
	subscribe() {},
});

const TestComponent = () => <div>Hello World</div>;
const TestComponentConnectWithMatchMedia = connectWithMatchMedia(
	TestComponent
);

const MATCH_MEDIA_FN_MOCK = mq => ({
	matches: false,
	addListener: jest.fn(),
	removeListener: jest.fn(),
});

describe('connectWithMatchMedia', () => {
	const mockStore = createFakeStore({
		config: { media: { isAtSmallUp: true } },
	});
	const connectWithMatchMedia = shallow(
		<TestComponentConnectWithMatchMedia />,
		{ context: { store: mockStore } }
	);
	it('exists', () => {
		expect(connectWithMatchMedia).toMatchSnapshot();
	});
});
