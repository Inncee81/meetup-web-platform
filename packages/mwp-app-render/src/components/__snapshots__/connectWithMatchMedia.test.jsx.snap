// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectWithMatchMedia exists 1`] = `
<ConnectWithMatchMedia(TestComponent)
  dispatch={[Function]}
  media={
    Object {
      "isAtSmallUp": true,
    }
  }
/>
`;
